library/std/src/sys/unix/process/process_unix.rs

```diff
@@ -436,6 +467,35 @@ impl Command {
             }
         }
 
+        // On QNX Neutrino, posix_spawnp can fail with EBADF in case "another thread might have opened
+        // or closed a file descriptor while the posix_spawn() was occurring".
+        // Documentation says "... or try calling posix_spawn() again". This is what we do here.
+        // See also http://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.lib_ref/topic/p/posix_spawn.html
+        #[cfg(target_os = "nto")]
+        unsafe fn retrying_libc_posix_spawnp(
+            pid: *mut pid_t,
+            file: *const c_char,
+            file_actions: *const posix_spawn_file_actions_t,
+            attrp: *const posix_spawnattr_t,
+            argv: *const *mut c_char,
+            envp: *const *mut c_char,
+        ) -> i32 {
+            // arbitrary number of retries, will retry for at least 10 seconds
+            let mut retries = 10_000;
+            loop {
+                match libc::posix_spawnp(pid, file, file_actions, attrp, argv, envp) {
+                    libc::EBADF if retries > 0 => {
+                        thread::sleep(Duration::from_millis(1));
+                        retries -= 1;
+                        continue;
+                    }
+                    r => {
+                        return r;
+                    }
+                }
+            }
```

Comment:
Same as with fork above.

Reply:
Unfortunately, changing this to a spawn() call won't fix the issue. Both spawn() and posix_spawn() follow similar code paths internally.  The issue occurs in a multi-threaded spawn from a process that attempts to dupe fd's that are bad or are in a state of flux.  We've fixed this behaviour in our new code, but it still exists for SDP 7.1 which is the target we're starting with. 

Although *possible* with some really poorly written code, this code path should not take seconds.  Even with a ton of thread contention, I would assume it would still be milliseconds, not seconds as FD handling will happen in short time increments in the kernel. 

Both Florian and myself were hesitant to add this code initially as we knew it would probably spark some resistance.  We were correct!  I view this code as an abstraction of the libc crate calls.  Different OS's call different APIs and handle the results differently.  Semantics aside, I think the random() example Florian provided is a good example of retrying a failed call with the expectation that it won't be infinite.  If we leave it up to the higher level code, I'm concerned users will not know why things are failing.  A call will simply error out on a QNX target randomly, but will succeed on other *nix distros.  Finding the root cause of an API failing may be challenging (we document EBADF on spawn, but once abstracted, finding the libc error and tracing it to our docs may be quite challenging).

I'm not against the view that the higher level code could/should be fixed to avoid this issue through locking/fd handling, but again there might be a bunch of existing code out there already that people will assume should just work as it originated on a non-QNX target.  That's my reasoning behind having this extra handling here.

We will of course defer to your expertise!  If you believe this shouldn't be handled here, we can look to fix the test server and we can work on solving issue of higher level code as they pop up in the future.  Looking forward to getting a QNX target upstreamed and apppreciate your feedback.

Reply:
@fritschy Just checking in as I noticed @flba-eb removed the loops.  My 2 cents would be to leave them, but my expertise lies outside of Rust, so looking for guidance.  Cheers.

Reply:
Culturally, Rust programmers tend to have a desire to handle their own errors, which is part of why you got such pushback. I don't think retrying again is out of the question, however, just with more restraint.